package/MDAnalysis/core/selection.py

```diff
@@ -217,6 +217,21 @@ def apply(self, group):
         return func(self, group)
     return apply
 
+
+def return_empty_on_empty_selection(func):
+    """
+    Decorator to return empty AtomGroups from the apply() function
+    if entire the selection is empty or a sub-expression is empty.
+    """
+    @functools.wraps(func)
+    def apply(self, group):
+        if len(group) == 0:
+            return group
+        if len(self.sel.apply(group)) == 0:
+            return group[[]]
+        return func(self, group)
```

Comment:
@richardjgowers will know efficient this is and whether the following suggestion is a good idea, but calling `self.sel.apply(group)` twice (once in the decorator, once in the function it's decorating) seems a little wasteful. Could you do `sel = self.sel.apply(group)` and pass `sel` in as an extra keyword argument? This would involve modifying the function signature of the relevant `apply`s, which might be a bad idea.

Reply:
I could definitely just pass `sel` in as an extra keyword. I don't like that it would make the function signature of `apply` inconsistent between classes though. I could also go back to implementing it in the functions, but that's very repetitive. This seems like largely a stylistic choice, so just let me know how I should proceed!

Reply:
Yeah I think Lily is right that this is neater but less efficient.  I'd say duplicate the code a few times

Reply:
Done. I let me know if there is anything else I should do before this gets merged.